clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/deploy/util/id/DcosSpinnakerAppId.groovy

```diff
@@ -0,0 +1,197 @@
+package com.netflix.spinnaker.clouddriver.dcos.deploy.util.id
+
+import static com.google.common.base.Strings.nullToEmpty
+
+import org.slf4j.LoggerFactory
+
+import com.netflix.frigga.Names
+
+/**
+ * Represents a hierarchical Spinnaker specific application identifier for DCOS.
+ * Structure - /account/region/app-stack-detail-sequence
+ */
+class DcosSpinnakerAppId {
+    private final static def LOGGER = LoggerFactory.getLogger(DcosSpinnakerAppId)
+    public final static def SAFE_REGION_SEPARATOR = "_"
+
+    private final def marathonPath
+
+    private DcosSpinnakerAppId(final MarathonPathId marathonPath) {
+        this.marathonPath = marathonPath
+    }
+
+    public String getAccount() {
+        marathonPath.first().get()
+    }
+
+    /**
+     * @return The canonical DC/OS "region" (a.k.a the full group path in which the marathon application lives)
+     *         including backslashes. This is returned as a relative path, meaning no preceeding backslash. Will never
+     *         be null.
+     *         <p/>
+     *         Deemed unsafe because various Spinnaker components have trouble with a region with backslashes.
+     *         <p/>
+     *         Ex: {@code foo/bar}
+     * @see #getSafeGroup()
+     */
+    public String getUnsafeGroup() {
+        marathonPath.tail().parent().relative().toString()
+    }
+
+    /**
+     * @return The "safe" DC/OS region (a.k.a the group in which the marathon application lives). This is returned as a
+     *         relative path, meaning no preceeding underscore. Will never be null.
+     *         <p/>
+     *         Deemed safe because backslashes are replaced with underscores.
+     *         <p/>
+     *         Ex: {@code acct_foo_bar}
+     * @see #getUnsafeGroup()
+     */
+    public String getSafeGroup() {
+        unsafeGroup.replaceAll(MarathonPathId.PART_SEPARATOR, SAFE_REGION_SEPARATOR)
+    }
+
+    public String getNamespace() {
+        marathonPath.parent().toString()
+    }
+
+    public Names getServerGroupName() {
+        Names.parseName(marathonPath.last().get())
+    }
+
+    @Override
+    public String toString() {
+        marathonPath.toString()
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o)
+            return true
+        if (o == null || getClass() != o.getClass())
```

Comment:
Can this be replaced with `if (!(o instanceof DcosSpinnakerAppId))`?
Or `if (o?.class != DcosSpinnakerAppId )`?